src/Core/src/WindowOverlay/WindowOverlay.Windows.cs

```diff
@@ -0,0 +1,144 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Microsoft.Graphics.Canvas.UI.Xaml;
+using Microsoft.Maui.Graphics;
+using Microsoft.Maui.Graphics.Win2D;
+using Microsoft.Maui.Handlers;
+using Microsoft.UI.Xaml;
+using Microsoft.UI.Xaml.Controls;
+using Windows.UI.Core;
+
+namespace Microsoft.Maui
+{
+	public partial class WindowOverlay : IWindowOverlay
+	{
+		private W2DGraphicsView? _graphicsView;
+		private bool disableUITouchEventPassthrough;
+		private Frame? _frame;
+		private RootPanel? _rootPanel;
+		private FrameworkElement? _nativeWindow;
+
+		/// <inheritdoc/>
+		public bool DisableUITouchEventPassthrough
+		{
+			get { return disableUITouchEventPassthrough; }
+			set
+			{
+				disableUITouchEventPassthrough = value;
+				if (_graphicsView != null)
+					_graphicsView.IsHitTestVisible = value;
+			}
+		}
+
+		/// <inheritdoc/>
+		public bool InitializeNativeLayer()
+		{
+			if (IsNativeViewInitialized)
+				return true;
+
+			if (Window?.Content == null)
+				return false;
+
+			_nativeWindow = Window.Content.GetNative(true);
+			if (_nativeWindow == null)
+				return false;
+			var handler = Window.Handler as WindowHandler;
+			if (handler == null || handler._rootPanel == null)
+				return false;
+
+			_rootPanel = handler._rootPanel;
+			// Capture when the frame is navigating.
+			// When it is, we will clear existing adorners.
+			if (_nativeWindow is Frame frame)
+			{
+				_frame = frame;
+				_frame.Navigating += FrameNavigating;
+			}
+
+			_graphicsView = new W2DGraphicsView() { Drawable = this };
+			if (_graphicsView == null)
+				return false;
+
+			_nativeWindow.Tapped += ViewTapped;
+			_nativeWindow.PointerMoved += PointerMoved;
+			_graphicsView.Tapped += ViewTapped;
+			_graphicsView.PointerMoved += PointerMoved;
+
+			_graphicsView.SetValue(Canvas.ZIndexProperty, 99);
```

Comment:
Some developers use really high numbers to make sure that their view is on top of everything. According to docs max value for Canvas.ZIndex is 1 million and since adorners will be last element added to the panel's Children it'll be always on top with this maximum value.

Also not sure if it's supported at the moment but what about PopupRoot XAML and adding elements to root children at runtime? Will invalidate cover this scenario?

Reply:
That's a good thought. IMO If a developer has their own instance of a WinUI panel and is changing the Z-Index, then whatever value the VisualDiagnostics overlay is set to doesn't matter at that point, since you're already beyond MAUI itself. 

As I see it right now, this API is tackling the common use cases for what MAUI does (A user building an app with the standard controls in XAML or code-behind), so we can build UI tooling and frameworks around it. If someone makes their own Window implementation and does new things with Panels/Android Views/UIViews and screws around with the layers or z-indexes, it could break this implementation... but you've already gone beyond what this tooling is for by that point.

If that becomes a common issue, then we could rethink this approach, but I think unless that happens I'm not sure tackling the edge cases will help if that makes sense.

As for ContentDialogs/PopupRoot, I found those don't provide the same level of access to the MAUI RootPanel then the approach I took, where I'm creating my own MAUI Graphics Layer and pushing it as a child view. We're looking for views that can stay on top of the app at all times, such as if you wanted to have something like a ToolBar or button that was always above your UI. ContentDialogs, if I remember right, don't allow for UI interaction under them unless you do some light hacking, and what I wrote should hopefully be more robust.